### PR TITLE
Iss 246 # Step-level retry

### PR DESCRIPTION
**Pull request by BartRobeyns**
Proposal for #246

### CODE REVIEW

File: core/src/main/java/org/jsmart/zerocode/core/runner/ZeroCodeMultiStepsScenarioRunnerImpl.java [hunk 1]
```diff
@@ -102,8 +102,19 @@ public synchronized boolean runScenario(ScenarioSpec scenario, RunNotifier notif
 
             for (Step thisStep : scenario.getSteps()) {
 
-                final int stepLoopTimes = thisStep.getLoop() == null ? 1 : thisStep.getLoop();
+                boolean retryTillSuccess = false;
+                int delay = 0;
+                int stepLoopTimes = 1;
```
Comment (officiallysameer) [inline core/src/main/java/org/jsmart/zerocode/core/runner/ZeroCodeMultiStepsScenarioRunnerImpl.java]: Please do not mix loop and retries
User a different variable e.g. maxRetryAttempts to store number of retries

File: core/src/main/java/org/jsmart/zerocode/core/runner/ZeroCodeMultiStepsScenarioRunnerImpl.java [hunk 2]
```diff
@@ -102,8 +102,19 @@ public synchronized boolean runScenario(ScenarioSpec scenario, RunNotifier notif
...
+                if ( thisStep.getLoop() != null ) {
+                    stepLoopTimes = thisStep.getLoop();
+                } else if ( thisStep.getRetry() != null ) {
+                    stepLoopTimes = thisStep.getRetry().getMax();
+                    delay = thisStep.getRetry().getDelay();
+                    retryTillSuccess = true;
+                }
+
                 for (int i = 0; i < stepLoopTimes; i++) {
+                    waitForDelay(delay);
```
Comment (officiallysameer) [inline core/src/main/java/org/jsmart/zerocode/core/runner/ZeroCodeMultiStepsScenarioRunnerImpl.java]: The delay should only be  used for retires(incase of failures), not to introduce delays for iteration of every loop

**Comment by authorjapps** [comment]
Thanks for the PR.

>Conflicting files
core/src/main/java/org/jsmart/zerocode/core/engine/mocker/RestEndPointMocker.java

Can you please sync/resolve these conflicts

**Comment by authorjapps** [comment]
Loop and retry can coexist in a step.

### "retry":
+ `retry` happens only if the assertions fail(as per the ticket)
+ `retry` doesn't happen for success scenarios

### "loop":
Continues for success scenarios.

Let's not override "loop" features with retry features.

You can take the `max retry count` to another variable and implement the ticket, please.

File: core/src/main/java/org/jsmart/zerocode/core/runner/ZeroCodeMultiStepsScenarioRunnerImpl.java (hunk 2 above)
Comment (BartRobeyns) [reply core/src/main/java/org/jsmart/zerocode/core/runner/ZeroCodeMultiStepsScenarioRunnerImpl.java]: The primary scenario I was considering was checking the result of an asynchronous operation, where an initial delay would be sensible.
But maybe that should be expressed outside of the retry-structure, in an 'initial-delay' attribute of the Step?
OK, I'll adapt the retry-delay.

**Comment by BartRobeyns** [comment]
If I understand @officiallysameer suggestion correctly, I like it: every iteration of the loop is retried max n-times.

More concrete:
`step: { loop: 3, retry: 5 }`
would always loop 3 times, and within each loop, the request is tried until it succeeds, maximum 5 times. So the maximum amount of requests would be 15, the minimum 3.

Does everyone agree with this?

**Comment by BartRobeyns** [comment]
Updated the PR to implement the 'retry-per-loop-iteration'

**Comment by BartRobeyns** [comment]
Hi guys, can you offer some direction on how to move forward with this PR? Thanks!

**Comment by authorjapps** [comment]
@officiallysameer is on it. He should be replying here soon(with any observations). Sorry @BartRobeyns for the delay!
Just validating nothing is going to break for existing users(_unintentionally_) and we have enough test coverage to take care of existing usecases...

**Comment by officiallysameer** [comment]
@BartRobeyns 
Apologies, was caught up in other things. Looking at it now.

**Comment by officiallysameer** [comment]
@BartRobeyns 

I have gone through the code changes and it looks mostly fine.
The tests however are not clear. Can you please explain more on how you are testing the new feature in your tests
for example, scenario 'Rest with Retry within loop Test'

**Comment by BartRobeyns** [comment]
@officiallysameer 
Do you need more documentation in the test itself (comments in the json-file) or an explanation here in the PR? 
Or would you like a different kind of test? I don't think the json-based tests are especially useful as unit-tests for this feature (and it's pretty hard to test all possible scenarios), but I don't see any other way to do it, without refactoring the runScenario-method in ZeroCodeMultiStepsScenarioRunnerImpl. Would love to make that one unit-testable, but that's not really in the scope of this PR.

**Comment by authorjapps** [comment]
> but I don't see any other way to do it, without refactoring the runScenario-method in ZeroCodeMultiStepsScenarioRunnerImpl. 

Fair one. Agree with you. This `Impl` has grown a bit larger than usual over time. 
Can you(when you get chance) please raise a tech-debt ticket with suggestions(to refactor/split) around it? Will keep this as a lower priority one. But we will try to pick it soonish.

**Comment by officiallysameer** [comment]
> @officiallysameer
> Do you need more documentation in the test itself (comments in the json-file) or an explanation here in the PR?
> Or would you like a different kind of test? I don't think the json-based tests are especially useful as unit-tests for this feature (and it's pretty hard to test all possible scenarios), but I don't see any other way to do it, without refactoring the runScenario-method in ZeroCodeMultiStepsScenarioRunnerImpl. Would love to make that one unit-testable, but that's not really in the scope of this PR.

Hi @BartRobeyns 

Please see the following points
1. The tests passes even after last retry assertion has failed. (tried with assertion status 400 in your test. The test retried correct number of times, but final result was success, which should be failure)
> Can you please fix this?

1. The test do not verify if the wiremocked endpoint was invoked expected number of times. Wiremock provides this functionality. Can you please include this to verify the invocation attempts? http://wiremock.org/docs/verifying/

1. The below log is not much meaningful. Can you improve it 
> LOGGER.info("continuing until success"); 
>  e.g. LOGGER.info("Retrying until success: Attempt number:  {}", attemptNumber);

**Comment by BartRobeyns** [comment]
@officiallysameer
 
>     1. The tests passes even after last retry assertion has failed. (tried with assertion status 400 in your test. The test retried correct number of times, but final result was success, which should be failure)

Oops, thanks for catching this. Fixed it.
>     2. The test do not verify if the wiremocked endpoint was invoked expected number of times. Wiremock provides this functionality. Can you please include this to verify the invocation attempts? http://wiremock.org/docs/verifying/

We can't interact with wiremock when running the integration-tests (the wiremock instance is constructed and destructed in each 'runScenario'). This would be achievable with a serious refactoring of ZeroCodeMultiStepsScenarioRunnerImpl, to make the logic itself better testable and provide better testing facilities for the scenario-based tests. As requested by @authorjapps, I'll try to write this down in a new ticket.

I did provide a different way of verifying the number of runs, based on the report-output, as it is done in ZeroCode..._FailedAssertionsTest. Please have a look at ZeroCodeMultiStepsScenarioRunnerImplRetryTest.
> 
>     3. The below log is not much meaningful. Can you improve it
> 
Fixed

**Comment by officiallysameer** [comment]
@BartRobeyns 
You can verify the happy case by adding the following step.
{
      "name": "GetDatesRequest",
      "url": "http://localhost:8888/__admin/requests/count",
      "operation": "POST",
      "request": {
        "body": {
          "url": "/retry/001",
          "method": "GET"
        }
      },
      "assertions": {
        "status": 200,
        "body" : {
          "count" : 1
        }
      }
    }

**Comment by BartRobeyns** [comment]
Thanks @officiallysameer, failed to see that one on the link you provided earlier.
But I did achieve the same using a different technique; is that ok for you or do you want it rewritten using the wiremock-counter? The downside to that approach is that two very similar things (counting requests in a successful scenario and counting requests in a failing scenario) would be expressed in totally different ways.

**Comment by BartRobeyns** [comment]
Holidays? :)

**Comment by authorjapps** [comment]
Sorry mate, yes, some of us are on holidays( with less access to the internet) and some are busy with other priority issues(in their limited time... with regular day jobs on). Thanks for understanding!
The PR is ok now. Thanks for your great work. We will take care of the merging.

But prior to that, we are changing some functionalities around the `loop`, as part of other tickets. That's why the merging has delayed!
We might need your help to change (if anything comes up) in this PR, but we will ping you when we reach there.

**Comment by BartRobeyns** [comment]
Great! Happy to help out if I can.

**Comment by authorjapps** [comment]
Folks, can you(all) approve this now?

**Pull request merged**